ggml/src/ggml-vulkan/vulkan-shaders/conv2d_mm.comp

```diff
@@ -0,0 +1,244 @@
+layout (binding = 0) readonly buffer A {A_TYPE knl_data[];};    // src0 - kernel:   [KW, KH, Cin, Cout] 
+layout (binding = 1) readonly buffer B {B_TYPE src_data[];};    // src1 - input:    [W, H, Cin, N] -- channel_first format
+layout (binding = 2) writeonly buffer D {D_TYPE dst_data[];};   // dst - result:    [OW, OH, Cout, N]
+
+layout (push_constant) uniform parameter {
+    // I/O channels, batch size
+    uint32_t Cout;
+    uint32_t Cin;
+    uint32_t N;
+    
+    // Tensor spatial sizes: kernel, input, output
+    uint32_t KW;
+    uint32_t KH;
+    uint32_t W;
+    uint32_t H;
+    uint32_t OW;
+    uint32_t OH;
+
+    // Parameters: stride, padding, dilation - 0=y, 1=x
+    uint32_t s0;
+    uint32_t s1;
+    uint32_t p0;
+    uint32_t p1;
+    uint32_t d0;
+    uint32_t d1;
+
+    // Strides in elements
+    uint32_t nb01;
+    uint32_t nb02;
+    uint32_t nb03;
+
+    uint32_t nb11;
+    uint32_t nb12;
+    uint32_t nb13;
+
+    uint32_t nb1;
+    uint32_t nb2;
+    uint32_t nb3;
+} p;
+
+#define WG_SIZE 256
+
+layout(local_size_x = WG_SIZE, local_size_y = 1, local_size_z = 1) in;
+
+uint32_t tid = gl_LocalInvocationID.x;
+const uint32_t bs = gl_WorkGroupSize.x;
+
+uint splitWork(uint work_size, uint block_size){
+    return (block_size + work_size -1) / block_size;
+}
+
+uint32_t K = p.Cout;
+uint32_t CRS = p.Cin*p.KH*p.KW;
+uint32_t NPQ = p.N*p.OH*p.OW;
+
+uint32_t n_elems_out = K*NPQ;
+
+// Blocktile sizes
+const uint32_t BS_K = 128;
+const uint32_t BS_CRS = 16;
+const uint32_t BS_NPQ = 128;
+
+// Number of blocktiles per input
+uint32_t NB_CRS = splitWork(CRS, BS_CRS);
+
+const uint32_t Ash_stride = BS_CRS+1;
+const uint32_t Bsh_stride = BS_NPQ+1;
+
+const uint32_t Ash_numel = BS_K*BS_CRS;
+const uint32_t Bsh_numel = BS_CRS*BS_NPQ;
+
+const uint32_t Ash_len = BS_K*Ash_stride;
+const uint32_t Bsh_len = BS_CRS*Bsh_stride;
+
+shared float Ash[Ash_len];  // K x CRS
+shared float Bsh[Bsh_len];  // CRS x NPQ
+
+// Threadtile sizes
+const uint32_t TS_K = 16;
+const uint32_t TS_NPQ = BS_K*BS_NPQ / WG_SIZE / TS_K;
+
+// Number of threadtiles per blocktile
+const uint32_t NT_K = BS_K / TS_K;
+const uint32_t NT_NPQ = BS_NPQ / TS_NPQ;
+
+float regA[TS_K];
+float regB[TS_NPQ];
+float regC[TS_K][TS_NPQ];
+
+/*
+Compute
+KxCRS @ CRSxNPQ = K x NPQ
+K=Cout
+C=Cin
+R,S=KH,KW
+P,Q=OH,OW
+*/
+
+uint32_t B_idx_K = gl_WorkGroupID.x;
+uint32_t B_idx_NPQ = gl_WorkGroupID.y;
+
+uint32_t T_y = tid / NT_NPQ;
+uint32_t T_x = tid % NT_NPQ;
+
+uint32_t Ar = tid / BS_CRS;
+uint32_t Ac = tid % BS_CRS;
+uint32_t ArpWg = WG_SIZE / BS_CRS;
+
+uint32_t Br = tid / BS_NPQ;
+uint32_t Bc = tid % BS_NPQ;
+uint32_t BrpWg = WG_SIZE / BS_NPQ;
+
+void initReg(){
+    for(uint32_t T_ly = 0; T_ly < TS_K; T_ly++){
+        for(uint32_t T_lx = 0; T_lx < TS_NPQ; T_lx++){
+            regC[T_ly][T_lx] = 0.0;
+        }
+    }
+}
+
+void outProdReg(){
+    for(uint32_t CRS_lidx = 0; CRS_lidx < BS_CRS; CRS_lidx++){
```

Comment:
Can you use coopmat here to do the outer products in parallel?

Reply:
If you ask if the alg supports coopmats -- then yes, we can add it later, but now I focus on achieving good enough flops relative to what can be achieved with the scalar matmul kernel.

Reply:
Subsequent PR will add this functionality.